### DISCUSSION TITLE
Angular Universal (SSR) does not have session

### DISCUSSION THREAD

**Discussion by rockcoolt** [ory/kratos#2816] [Q&A] [answered]
Hello, 
I have a question: I implemented Kratos flow on an Angular SSR SPA.

I have a browser valid session cookie; 
- When i disable the javascript to call the angular SSR version, Kratos returns to me a 401, it feels like there is no session (_maybe Kratos doesn't set the cookie!_)
BUT IF i active again the Javascrit, my session is here. 

Is it normal ? 

I'm wondering why Kratos doesn't put the cookie for SSR SPA ?

**Comment by vinckr** [comment, marked as answer]
Hello @rockcoolt 
See this tip: https://www.ory.sh/docs/kratos/debug/csrf#accessing-ory-kratos-apis-from-a-server-side-application 

> When implementing a server-side application (for example in PHP, Java, Go, Node.js, ...) make sure to include the Cookie header from the client when fetching the self-service flows (for example GET /self-service/login/flows?id=...).
> 
> Without forwarding the Cookie HTTP Header it won't be possible to fetch the flow due to a security error. This protection prevents accidental leak of personal information when users copy/paste, for example, the login URL.

Also see this blogpost, it goes a bit into the SPA/SSR flows: https://www.ory.sh/auth-and-modern-software/ 

Let me know if that helped!

**Reply by rockcoolt** [reply]
Yes you are right!
I found a solution: In angular **interceptor** i did that:

```
constructor(@Optional() @Inject(REQUEST) private request: any) {}
.....

intercept(req: HttpRequest<SafeAny>, next: HttpHandler): Observable<HttpEvent<SafeAny>> {
 ......
 
  let newReq = req.clone({
        url,
        withCredentials: true
      });

      if (this.request) {
        // If optional request is provided, we are server side
        let headers = new HttpHeaders();
        headers = headers.set('cookie', this.request.headers.cookie);
        newReq = req.clone({
          headers: headers,
          withCredentials: true,
          url
        });
      }
      return next.handle(newReq);
 } 
```

if you are using the SDK you have to put the cookie in parameter:

```toSession(undefined, this.request?.headers?.cookie)```

**Reply by vinckr** [reply]
Awesome, glad it got resolved quickly 🙏
